File: tel/otlp.py
"""Decoding of OTLP/JSON records as written by the collector's file exporter.

The file exporter writes one JSON object per line, each holding a whole
export request (``resourceLogs`` / ``resourceMetrics`` / ``resourceSpans``).
These helpers flatten those into individual records while keeping the
resource and scope context attached.
"""
from __future__ import annotations

import datetime as _dt
from typing import Any, Iterator


def any_value(v: Any) -> Any:
    """Convert an OTLP AnyValue into a plain Python value."""
    if not isinstance(v, dict):
        return v
    if "stringValue" in v:
        return v["stringValue"]
    if "intValue" in v:
        try:
            return int(v["intValue"])
        except (TypeError, ValueError):
            return v["intValue"]
    if "doubleValue" in v:
        try:
            return float(v["doubleValue"])
        except (TypeError, ValueError):
            return v["doubleValue"]
    if "boolValue" in v:
        return bool(v["boolValue"])
    if "arrayValue" in v:
        return [any_value(x) for x in v["arrayValue"].get("values", [])]
    if "kvlistValue" in v:
        return attrs(v["kvlistValue"].get("values", []))
    if "bytesValue" in v:
        return v["bytesValue"]
    return None


def attrs(kvs: Any) -> dict:
    """Convert an OTLP KeyValue list into a dict."""
    out: dict = {}
    if not kvs:
        return out
    for kv in kvs:
        if isinstance(kv, dict) and "key" in kv:
            out[kv["key"]] = any_value(kv.get("value"))
    return out


def ns_to_iso(ns: Any) -> str | None:
    """Nanosecond epoch (string or int) to an ISO-8601 UTC timestamp."""
    if ns in (None, "", 0, "0"):
        return None
    try:
        n = int(ns)
    except (TypeError, ValueError):
        return None
    if n <= 0:
        return None
    return (
        _dt.datetime.fromtimestamp(n / 1e9, tz=_dt.timezone.utc)
        .isoformat(timespec="microseconds")
        .replace("+00:00", "Z")
    )


def to_ns(ns: Any) -> int | None:
    try:
        n = int(ns)
        return n if n > 0 else None
    except (TypeError, ValueError):
        return None


def iter_logs(payload: dict) -> Iterator[dict]:
    """Yield one dict per log record, with resource/scope context merged in."""
    for rl in payload.get("resourceLogs", []) or []:
        resource = attrs((rl.get("resource") or {}).get("attributes"))
        for sl in rl.get("scopeLogs", []) or []:
            scope = (sl.get("scope") or {}).get("name")
            for rec in sl.get("logRecords", []) or []:
                yield {
                    "resource": resource,
                    "scope": scope,
                    "record": rec,
                    "attributes": attrs(rec.get("attributes")),
                    "body": any_value(rec.get("body")),
                    "ts_ns": to_ns(rec.get("timeUnixNano"))
                    or to_ns(rec.get("observedTimeUnixNano")),
                    "trace_id": rec.get("traceId") or None,
                    "span_id": rec.get("spanId") or None,
                    "severity": rec.get("severityText") or None,
                }


_POINT_KINDS = ("sum", "gauge", "histogram", "exponentialHistogram", "summary")


def iter_metric_points(payload: dict) -> Iterator[dict]:
    """Yield one dict per metric data point."""
    for rm in payload.get("resourceMetrics", []) or []:
        resource = attrs((rm.get("resource") or {}).get("attributes"))
        for sm in rm.get("scopeMetrics", []) or []:
            scope = (sm.get("scope") or {}).get("name")
            for metric in sm.get("metrics", []) or []:
                name = metric.get("name")
                unit = metric.get("unit")
                for kind in _POINT_KINDS:
                    block = metric.get(kind)
                    if not block:
                        continue
                    for dp in block.get("dataPoints", []) or []:
                        if "asInt" in dp:
                            value = float(dp["asInt"])
                        elif "asDouble" in dp:
                            value = float(dp["asDouble"])
                        elif "sum" in dp:  # histogram
                            value = float(dp["sum"])
                        else:
                            value = None
                        yield {
                            "resource": resource,
                            "scope": scope,
                            "metric_name": name,
                            "unit": unit,
                            "kind": kind,
                            "value": value,
                            "attributes": attrs(dp.get("attributes")),
                            "ts_ns": to_ns(dp.get("timeUnixNano"))
                            or to_ns(dp.get("startTimeUnixNano")),
                            "point": dp,
                        }


def iter_spans(payload: dict) -> Iterator[dict]:
    """Yield one dict per span."""
    for rs in payload.get("resourceSpans", []) or []:
        resource = attrs((rs.get("resource") or {}).get("attributes"))
        for ss in rs.get("scopeSpans", []) or []:
            scope = (ss.get("scope") or {}).get("name")
            for span in ss.get("spans", []) or []:
                start = to_ns(span.get("startTimeUnixNano"))
                end = to_ns(span.get("endTimeUnixNano"))
                yield {
                    "resource": resource,
                    "scope": scope,
                    "span": span,
                    "name": span.get("name"),
                    "trace_id": span.get("traceId"),
                    "span_id": span.get("spanId"),
                    "parent_span_id": span.get("parentSpanId") or None,
                    "attributes": attrs(span.get("attributes")),
                    "start_ns": start,
                    "end_ns": end,
                    "duration_ms": ((end - start) / 1e6) if (start and end) else None,
                    "status": (span.get("status") or {}).get("code"),
                }


def signal_of(payload: dict) -> str | None:
    if "resourceLogs" in payload:
        return "logs"
    if "resourceMetrics" in payload:
        return "metrics"
    if "resourceSpans" in payload:
        return "traces"
    return None

File: src/otlp.ts
/** Decoding of OTLP/JSON records as written by the receiver.
 *
 * Each line of a raw file is a whole export request (`resourceLogs` /
 * `resourceMetrics` / `resourceSpans`). These helpers flatten those into
 * individual records while keeping the resource and scope context attached.
 */
import { nsToIsoString } from './util/text.js'

export type Attrs = Record<string, unknown>

/** Convert an OTLP AnyValue into a plain value. */
export function anyValue(v: unknown): unknown {
  if (v === null || typeof v !== 'object' || Array.isArray(v)) return v
  const obj = v as Record<string, any>
  if ('stringValue' in obj) return obj.stringValue
  if ('intValue' in obj) {
    const parsed = Number.parseInt(String(obj.intValue), 10)
    return Number.isNaN(parsed) ? obj.intValue : parsed
  }
  if ('doubleValue' in obj) {
    const parsed = Number.parseFloat(String(obj.doubleValue))
    return Number.isNaN(parsed) ? obj.doubleValue : parsed
  }
  if ('boolValue' in obj) return Boolean(obj.boolValue)
  if ('arrayValue' in obj) return (obj.arrayValue?.values ?? []).map(anyValue)
  if ('kvlistValue' in obj) return attrs(obj.kvlistValue?.values ?? [])
  if ('bytesValue' in obj) return obj.bytesValue
  return null
}

/** Convert an OTLP KeyValue list into an object. */
export function attrs(kvs: unknown): Attrs {
  const out: Attrs = {}
  if (!kvs || !Array.isArray(kvs)) return out
  for (const kv of kvs) {
    if (kv && typeof kv === 'object' && 'key' in kv) {
      out[(kv as any).key] = anyValue((kv as any).value)
    }
  }
  return out
}

/** Nanosecond epoch (string or number) as an exact integer, or null. */
export function toNs(value: unknown): bigint | null {
  if (value === null || value === undefined || value === '') return null
  try {
    const n = typeof value === 'bigint' ? value : BigInt(String(value).trim())
    return n > 0n ? n : null
  } catch {
    return null
  }
}

/** Nanosecond epoch to an ISO-8601 UTC timestamp. */
export function nsToIso(value: unknown): string | null {
  const n = toNs(value)
  return n === null ? null : nsToIsoString(n)
}

export interface LogRecord {
  resource: Attrs
  scope: string | null
  record: Record<string, any>
  attributes: Attrs
  body: unknown
  ts_ns: bigint | null
  trace_id: string | null
  span_id: string | null
  severity: string | null
}

/** One entry per log record, with resource/scope context merged in. */
export function* iterLogs(payload: Record<string, any>): Generator<LogRecord> {
  for (const rl of payload.resourceLogs ?? []) {
    const resource = attrs(rl?.resource?.attributes)
    for (const sl of rl?.scopeLogs ?? []) {
      const scope = sl?.scope?.name ?? null
      for (const rec of sl?.logRecords ?? []) {
        yield {
          resource,
          scope,
          record: rec,
          attributes: attrs(rec?.attributes),
          body: anyValue(rec?.body),
          ts_ns: toNs(rec?.timeUnixNano) ?? toNs(rec?.observedTimeUnixNano),
          trace_id: rec?.traceId || null,
          span_id: rec?.spanId || null,
          severity: rec?.severityText || null,
        }
      }
    }
  }
}

const POINT_KINDS = [
  'sum',
  'gauge',
  'histogram',
  'exponentialHistogram',
  'summary',
] as const

export interface MetricPoint {
  resource: Attrs
  scope: string | null
  metric_name: string | null
  unit: string | null
  kind: string
  value: number | null
  attributes: Attrs
  ts_ns: bigint | null
  point: Record<string, any>
}

/** One entry per metric data point. */
export function* iterMetricPoints(
  payload: Record<string, any>,
): Generator<MetricPoint> {
  for (const rm of payload.resourceMetrics ?? []) {
    const resource = attrs(rm?.resource?.attributes)
    for (const sm of rm?.scopeMetrics ?? []) {
      const scope = sm?.scope?.name ?? null
      for (const metric of sm?.metrics ?? []) {
        const name = metric?.name ?? null
        const unit = metric?.unit ?? null
        for (const kind of POINT_KINDS) {
          const block = metric?.[kind]
          if (!block) continue
          for (const dp of block.dataPoints ?? []) {
            let value: number | null
            if ('asInt' in dp) value = Number(dp.asInt)
            else if ('asDouble' in dp) value = Number(dp.asDouble)
            else if ('sum' in dp)
              value = Number(dp.sum) // histogram
            else value = null
            yield {
              resource,
              scope,
              metric_name: name,
              unit,
              kind,
              value,
              attributes: attrs(dp?.attributes),
              ts_ns: toNs(dp?.timeUnixNano) ?? toNs(dp?.startTimeUnixNano),
              point: dp,
            }
          }
        }
      }
    }
  }
}

export interface SpanRecord {
  resource: Attrs
  scope: string | null
  span: Record<string, any>
  name: string | null
  trace_id: string | null
  span_id: string | null
  parent_span_id: string | null
  attributes: Attrs
  start_ns: bigint | null
  end_ns: bigint | null
  duration_ms: number | null
  status: unknown
}

/** One entry per span. */
export function* iterSpans(
  payload: Record<string, any>,
): Generator<SpanRecord> {
  for (const rs of payload.resourceSpans ?? []) {
    const resource = attrs(rs?.resource?.attributes)
    for (const ss of rs?.scopeSpans ?? []) {
      const scope = ss?.scope?.name ?? null
      for (const span of ss?.spans ?? []) {
        const start = toNs(span?.startTimeUnixNano)
        const end = toNs(span?.endTimeUnixNano)
        yield {
          resource,
          scope,
          span,
          name: span?.name ?? null,
          trace_id: span?.traceId ?? null,
          span_id: span?.spanId ?? null,
          parent_span_id: span?.parentSpanId || null,
          attributes: attrs(span?.attributes),
          start_ns: start,
          end_ns: end,
          duration_ms:
            start !== null && end !== null ? Number(end - start) / 1e6 : null,
          status: span?.status?.code,
        }
      }
    }
  }
}
